Tokenized noreply address and check for + in emails

hackerkid opened the pull request.

timabbott commented:
A few comments on the FromAddress part of this branch:
* I did a `git grep FromAddress.NOREPLY`, and it looks like we left the digest/missed message email code paths as-is.  I think that's fine, but I'd like to think about whether there's a way to future-proof this (e.g. having `FromAddress.NOREPLY_CONFIRMATION` be a special value that we set outside `build_email`, and have `build_email` do the rewriting).  The reason that might be a good idea is that it feels easier to document what the various options are to prevent accidental regressions here.
* `manage.py send_test_email` should use the confirmation style, since that's what's relevant for email server whitelist issues.
* We should probably update `docs/production/email.html` to mention that Zulip sends email using some randomly generated from addresses starting with `noreply+`.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -2164,6 +2190,17 @@ def test_failed_signup_due_to_disposable_email(self) -> None:
+        request.session = {}  # type: ignore
+        email = 'iago+label@zulip.com'
+        form = HomepageForm({'email': email}, realm=realm)
+        self.assertIn("Email aliases that make use of + are not allowed.", form.errors['email'][0])
```
We should have a test that succeeds without `restricted_to_domain`, right?

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -306,6 +306,15 @@ def test_create_user_backend(self) -> None:
         result = self.client_post("/json/users", valid_params)
         self.assert_json_error(result, "Disposable email addresses are not allowed in this organization")
 
+        # Don't allow user to sign up with email aliases based on + when realm
```
This should say "Don't allow creating a user with + in their email address when ..."

timabbott commented on zerver/views/users.py:
```diff
@@ -425,6 +426,8 @@ def create_user_backend(request: HttpRequest, user_profile: UserProfile,
     except DisposableEmailError:
         return json_error(_("Disposable email addresses are not allowed in this organization"))
+    except EmailAliasNotAllowedForRealm:
+        return json_error(_("Email aliases that make use of + are not allowed."))
```
We need a better error message for this.  Maybe: "Email addresses containing `+` are not allowed in this organization."

@rishig FYI

rishig replied:
lgtm.
From a code reading perspective, it's a bit confusing that the error is EmailAliasNotAllowedForRealm rather than EmailContainsPlusError or something, but didn't spend much time reading the rest of the code.

showell commented on zerver/lib/send_email.py:
```diff
@@ -25,6 +26,12 @@ class FromAddress:
     NOREPLY = parseaddr(settings.NOREPLY_EMAIL_ADDRESS)[1]
 
+    @staticmethod
+    def tokenized_no_reply_address() -> str:
```
Can we add a couple tests that directly use this method?  I think the tests would be helpful not only to prevent regressions but also to illustrate what this does.  It's also kind of strange for a `tokenized_foo` message to return a `str`; I'd normally expect a "tokenized" thing to bee a list of strings.  So, it's possible the name here is a bit misleading, although it's also fine to just have a comment explain what it's doiing.

showell commented:
@hackerkid I added one comment.  Do you think this is ready for another round of review (i.e. have you addressed all of Tim's comments?)?

It can also be helpful on these types of PRs to call out that most of the commits are just small variations on the same pattern (as opposed to a typical long PR with lots of different commits).  It just helps us know that it's a possibly easier review than it appears at first glance.

showell commented on zerver/tests/test_email_change.py:
```diff
@@ -103,7 +105,8 @@ def test_end_to_end_flow(self) -> None:
         self.assertIn('Zulip Account Security', from_email)
-        self.assertIn(FromAddress.NOREPLY, email_message.from_email)
+        tokenized_no_reply_email = parseaddr(email_message.from_email)[1]
+        assert(re.search(self.TOKENIZED_NOREPLY_REGEX, tokenized_no_reply_email))
```
I think we prefer `self.assertTrue` to `assert` where we're making assertions about whether the real code works.  I like `assert()` for cases where we're asserting some invariant about the test itself, which doesn't seem to be the case here.

hackerkid commented:
@showell With the commit, I just added I have addressed all the comments so far left by Tim and Rishi. Will update the PR to address the comments you left.

showell commented:
Thanks for the update!

hackerkid commented:
@showell This is ready for a review.

timabbott commented:
I just pushed an updated version of this PR cleaning up the doc commit at the end.  

We should change the format to be `noreply+...`, not `noreply-...`, and otherwise lgtm.

rishig commented:
fwiw, other products do `noreply-`

timabbott commented:
I guess that may be safer.  Fine, will just stick with this as-is.
